src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx

```diff
@@ -9,34 +16,127 @@ export const lessonSections = [
   "review",
 ] as const;
 
-export const lessonReviewSections = [
+export const allLessonReviewSections = [
   "intro",
   "starter-quiz",
   "video",
   "exit-quiz",
 ] as const;
 
 export type LessonSection = (typeof lessonSections)[number];
+export type LessonReviewSection = (typeof allLessonReviewSections)[number];
 
 export const isLessonSection = (
   currentSection: string,
 ): currentSection is LessonSection => {
   return lessonSections.includes(currentSection as LessonSection);
 };
 
-export type LessonSectionResult = { grade: number; numQuestions: number };
+export function isLessonReviewSection(
+  section: string,
+): section is LessonReviewSection {
+  return lessonSections.includes(section as LessonReviewSection);
+}
+
+type QuizResult = { grade: number; numQuestions: number };
+type LessonSectionState = {
+  isComplete: boolean;
+} & Partial<QuizResult>;
+type LessonEngineAction =
+  | {
+      type: "setCurrentSection";
+      section: LessonSection;
+    }
+  | {
+      type: "completeSection";
+      section: LessonReviewSection;
+    }
+  | {
+      type: "updateQuizResult";
+      result: QuizResult;
+    }
+  | {
+      type: "proceedToNextSection";
+    };
+
+type LessonEngineState = {
+  currentSection: LessonSection;
+  lessonReviewSections: Readonly<LessonReviewSection[]>;
+  sections: Partial<Record<LessonReviewSection, LessonSectionState>>;
+};
 
-type LessonSectionResults = Partial<Record<LessonSection, LessonSectionResult>>;
+const lessonEngineReducer: Reducer<LessonEngineState, LessonEngineAction> = (
+  currentState,
+  action,
+) => {
+  switch (action.type) {
+    case "setCurrentSection":
+      return { ...currentState, currentSection: action.section };
+    case "completeSection": {
+      const newState = {
+        ...currentState,
+        sections: {
+          ...currentState.sections,
+          [action.section]: {
+            ...currentState.sections[action.section],
+            isComplete: true,
+          },
+        },
+      };
```

Comment:
These patterns of immutability can get a little tiring on the eye as levels of nesting grow. As that happens things can be further decomposed into sub-reducers. Also libraries like [Immer](https://immerjs.github.io/immer/) exist to make these things easier to follow.

For now this is a pretty trivial case.